docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET http://extension-name.com:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+- **Example 2**:
+
+If a backend provides an API under the `/apiv1/metrics` endpoint, Argo CD
+should be able to invoke it such as:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension/apiv1/metrics/123
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET http://extension-name.com:8080/apiv1/metrics/123
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+- **Example 3**:
+
+In this use-case we have one Argo CD instance connected with different
+clusters. There is a requirement defining that every extension instance
+needs to be deployed in each of the target clusters. To address this
+use-case there is a need to configure multiple backend URLs for the
+same extension (one for each cluster). For doing so, the following
+configuration should be possible:
+
+```yaml
+extension.config: |
+  extensions:
+    - name: some-extension
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        services:
+          - url: http://extension-name.com:8080
+            clusterName: kubernetes.local
+          - url: https://extension-name.ppd.cluster.k8s.local:8080
+            clusterName: admins@ppd.cluster.k8s.local
+```
+
+Note that there is an URL configuration per cluster name. The cluster
+name is extracted from the Argo CD cluster secret and must match the
+field `data.name`. In this case the UI must send a header with the
+application name so API Server can check in which cluster it should
+get the backend URL from. This will be done by inspecting the
+Application destination configuration to find the proper cluster
+name.
+
+The diagram below shows how Argo CD UI could send the request with
+the additional header to get the proxy forwarding it to the proper
+cluster:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │ HEADER: "ARGOCD-APPLICATION-NAME: ppd-application
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET https://extension-name.ppd.cluster.k8s.local:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+##### Considerations
+
+- The `idleConnTimeout` can be used to avoid accumulating too many
+  goroutines waiting slow for extensions. In this case a proper timeout
+  error (408) should be returned to the browser.
+- Headers, scheme, http verb and request body are forwarded as it is
+  received by the API server to the backend service.
```

Comment:
Are you sure we should forward the Argo CD auth token? Seems like an unnecessary risk, as the described use cases do not communicate _back to_ the Argo CD API and therefore don't need the token.

Reply:
updated